### PR TITLE
Warn on nullability mismatch of constructed types and arrays when converting

### PR DESCRIPTION
**Pull request by cston**

### CODE REVIEW

**Comment by cston** [comment]
@dotnet/roslyn-compiler, @AlekseyTs please review.

File: src/Compilers/CSharp/Portable/Binder/Semantics/Conversions/ConversionsBase.cs [hunk 1]
```diff
@@ -1324,7 +1331,24 @@ public static bool HasIdentityConversion(TypeSymbol type1, TypeSymbol type2)
...
+            if (!type1.Equals(type2, TypeCompareKind.AllIgnoreOptions))
+            {
+                return false;
+            }
+
+            if (!type1.Equals(type2, TypeCompareKind.AllIgnoreOptions | TypeCompareKind.CompareNullableModifiersForReferenceTypes | TypeCompareKind.UnknownNullableModifierMatchesAny))
+            {
+                if (useSiteDiagnostics == null)
+                {
+                    useSiteDiagnostics = new HashSet<DiagnosticInfo>();
```
Comment (VSadov) [inline src/Compilers/CSharp/Portable/Binder/Semantics/Conversions/ConversionsBase.cs]: do we want to call this "useSiteDiagnostics" ?  
I think  useSiteDiagnostics is used to report diagnostics on broken/missing symbols. This seems to be just a regular diagnostics related to a conversion. (not attached to a symbol and reported at use site).

Comment (cston) [reply src/Compilers/CSharp/Portable/Binder/Semantics/Conversions/ConversionsBase.cs]: I agree, this should probably not be `useSiteDiagnostics`. This was a simple approach to allow returning the nullability warnings from conversion methods using an existing mechanism rather than changing the signatures of many methods. I will add a `// PROTOTYPE` comment to revisit this approach later.

File: src/Compilers/CSharp/Portable/Binder/Semantics/Conversions/ConversionsBase.cs [hunk 2]
```diff
@@ -1312,6 +1313,12 @@ internal bool HasCallerInfoStringConversion(TypeSymbol destination, ref HashSet<
 
         public static bool HasIdentityConversion(TypeSymbol type1, TypeSymbol type2)
         {
+            HashSet<DiagnosticInfo> useSiteDiagnostics = null;
+            return HasIdentityConversion(type1, type2, ref useSiteDiagnostics);
```
Comment (VSadov) [inline src/Compilers/CSharp/Portable/Binder/Semantics/Conversions/ConversionsBase.cs]: HasIdentityConversion is called extremely often in the compiler with "true" result. It is the first conversion that we check for since it is the most common.
We are deferring to a much heavier method that would call Equals two times in order to possibly create a HashSet with diagnostics that we will drop here anyways.

Can we just  'return type1.Equals(type2, TypeCompareKind.AllIgnoreOptions);'  here?

If we want the one helper to call another, perhaps the heavy one should call the simpler one?

File: src/Compilers/CSharp/Portable/Binder/Binder_Operators.cs [hunk 3]
```diff
@@ -3514,87 +3517,64 @@ private BoundExpression BindConditionalOperator(ConditionalExpressionSyntax node
...
-                if ((object)trueType == null)
+            if ((object)bestType == null)
+            {
+                // CONSIDER: Dev10 suppresses ERR_InvalidQM unless the following is true for both trueType and falseType
+                // (!T->type->IsErrorType() || T->type->AsErrorType()->HasTypeParent() || T->type->AsErrorType()->HasNSParent())
+                if (hadMultipleCandidates)
                 {
-                    // If trueExpr and falseExpr both have type null, then we don't have any symbols
-                    // to pass to a SymbolDistinguisher (which ERR_InvalidQM would usually require).
-                    diagnostics.Add(ErrorCode.ERR_InvalidQM, node.Location, trueExpr.Display, falseExpr.Display);
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/Binder_Operators.cs]: >diagnostics.Add(ErrorCode.ERR_InvalidQM, node.Location, trueExpr.Display, falseExpr.Display); [](start = 20, length = 93)

Are we still reporting this error for the same scenario?

File: src/Compilers/CSharp/Portable/Binder/Semantics/Conversions/ConversionsBase.cs [hunk 4]
```diff
@@ -807,10 +807,11 @@ private Conversion ClassifyImplicitBuiltInConversionFromExpression(BoundExpressi
                     break;
 
                 case BoundKind.SuppressNullableWarningExpression:
-                    var notNullConversion = ClassifyImplicitBuiltInConversionFromExpression(((BoundSuppressNullableWarningExpression)sourceExpression).Expression, source, destination, ref useSiteDiagnostics);
-                    if (notNullConversion.Exists)
+                    var innerExpression = ((BoundSuppressNullableWarningExpression)sourceExpression).Expression;
+                    var innerConversion = ClassifyImplicitBuiltInConversionFromExpression(innerExpression, innerExpression.Type, destination, ref useSiteDiagnostics);
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/Semantics/Conversions/ConversionsBase.cs]: >var innerConversion = ClassifyImplicitBuiltInConversionFromExpression(innerExpression, innerExpression.Type, destination, ref useSiteDiagnostics); [](start = 20, length = 146)

Should something be done to suppress the warnings?

File: src/Compilers/CSharp/Portable/Binder/Semantics/Conversions/ConversionsBase.cs [hunk 5]
```diff
@@ -1324,7 +1331,24 @@ public static bool HasIdentityConversion(TypeSymbol type1, TypeSymbol type2)
...
+            }
+
+            if (!type1.Equals(type2, TypeCompareKind.AllIgnoreOptions | TypeCompareKind.CompareNullableModifiersForReferenceTypes | TypeCompareKind.UnknownNullableModifierMatchesAny))
+            {
+                if (useSiteDiagnostics == null)
+                {
+                    useSiteDiagnostics = new HashSet<DiagnosticInfo>();
+                }
+                // PROTOTYPE(NullableReferenceTypes): Need general WRN_NullabilityMismatch warning.
+                useSiteDiagnostics.Add(new CSDiagnosticInfo(ErrorCode.WRN_NullabilityMismatchInAssignment, type1, type2));
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/Semantics/Conversions/ConversionsBase.cs]: >useSiteDiagnostics.Add(new CSDiagnosticInfo(ErrorCode.WRN_NullabilityMismatchInAssignment, type1, type2)); [](start = 16, length = 106)

Is it possible that we could end up with false positives? For example, we tried to check the identity conversion, but this wasn't used for final determination. We could even detect that the final conversion doesn't exist.

File: src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/OverloadResolution.cs [hunk 6]
```diff
@@ -3138,7 +3138,7 @@ private RefKind GetEffectiveParameterRefKind(ParameterSymbol parameter, RefKind
                 return conversion;
             }
 
-            if ((object)argType != null && Conversions.HasIdentityConversion(argType, parameterType))
+            if ((object)argType != null && Conversions.HasIdentityConversion(argType, parameterType, ref useSiteDiagnostics))
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/OverloadResolution.cs]: >ref useSiteDiagnostics [](start = 101, length = 22)

Can end up reporting a warning that is not related to the best result chosen?

File: src/Compilers/CSharp/Portable/Lowering/DiagnosticsPass_ExpressionTrees.cs [hunk 7]
```diff
@@ -4,7 +4,6 @@
 using System.Diagnostics;
 using Microsoft.CodeAnalysis.CSharp.Symbols;
 using Microsoft.CodeAnalysis.CSharp.Syntax;
-using Roslyn.Utilities;
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Lowering/DiagnosticsPass_ExpressionTrees.cs]: It looks like changes in this file are not related to the goal of this PR. Please revert and submit as a separate PR.

**Comment by AlekseyTs** [comment]
> {

Please exclude this file from this PR.

---
Refers to: src/Compilers/CSharp/Test/Semantic/Properties/launchSettings.json:1 in 799007d. [](commit_id = 799007d1ac4fd1b627e512103e86530bb5fd302f, deletion_comment = False)

File: src/Compilers/CSharp/Test/Semantic/Semantics/StaticNullChecking.cs [hunk 8]
```diff
@@ -3745,18 +3745,21 @@ class CL0<T>
 " }, parseOptions: TestOptions.Regular.WithNullCheckingFeature());
 
             c.VerifyDiagnostics(
-                // (12,16): warning CS8620: Nullability of reference types in argument of type 'CL0<string?>' doesn't match target type 'CL0<string>' for parameter 'x' in 'void C.M1(ref CL0<string> x)'.
+                // (12,9): warning CS8619: Nullability of reference types in value of type 'CL0<string?>' doesn't match target type 'CL0<string>'.
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/Semantics/StaticNullChecking.cs]: >// (12,9): warning CS8619: Nullability of reference types in value of type 'CL0<string?>' doesn't match target type 'CL0<string>'. [](start = 16, length = 130)

This looks like duplicate of the warning below.

Comment (AlekseyTs) [reply src/Compilers/CSharp/Test/Semantic/Semantics/StaticNullChecking.cs]: I see that this warning was reported before the change, but at a different location.

---
In reply to: [136656548](https://github.com/dotnet/roslyn/pull/21825#discussion_r136656548) [](ancestors = 136656548)

File: src/Compilers/CSharp/Test/Semantic/Semantics/StaticNullChecking.cs [hunk 9]
```diff
@@ -3745,18 +3745,21 @@ class CL0<T>
...
-                // (12,16): warning CS8619: Nullability of reference types in value of type 'CL0<string>' doesn't match target type 'CL0<string?>'.
+                Diagnostic(ErrorCode.WRN_NullabilityMismatchInAssignment, "M1").WithArguments("CL0<string?>", "CL0<string>").WithLocation(12, 9),
+                // (12,16): warning CS8620: Nullability of reference types in argument of type 'CL0<string?>' doesn't match target type 'CL0<string>' for parameter 'x' in 'void C.M1(ref CL0<string> x)'.
                 //         M1(ref x1);
-                 Diagnostic(ErrorCode.WRN_NullabilityMismatchInAssignment, "x1").WithArguments("CL0<string>", "CL0<string?>").WithLocation(12, 16),
-                // (19,16): warning CS8619: Nullability of reference types in value of type 'CL0<string?>' doesn't match target type 'CL0<string>'.
+                Diagnostic(ErrorCode.WRN_NullabilityMismatchInArgument, "x1").WithArguments("CL0<string?>", "CL0<string>", "x", "void C.M1(ref CL0<string> x)").WithLocation(12, 16),
+                // (19,9): warning CS8619: Nullability of reference types in value of type 'CL0<string>' doesn't match target type 'CL0<string?>'.
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/Semantics/StaticNullChecking.cs]: >warning CS8619: Nullability of reference types in value of type 'CL0<string>' doesn't match target type 'CL0<string?>'. [](start = 27, length = 119)

The previous error was probably more accurate because it was reflecting the direction in which values flowing.

File: src/Compilers/CSharp/Test/Semantic/Semantics/StaticNullChecking.cs [hunk 10]
```diff
@@ -3745,18 +3745,21 @@ class CL0<T>
...
                 //         M2(out x2);
-                 Diagnostic(ErrorCode.WRN_NullabilityMismatchInAssignment, "x2").WithArguments("CL0<string?>", "CL0<string>").WithLocation(19, 16),
+                Diagnostic(ErrorCode.WRN_NullabilityMismatchInAssignment, "M2").WithArguments("CL0<string>", "CL0<string?>").WithLocation(19, 9),
+                // (26,9): warning CS8619: Nullability of reference types in value of type 'CL0<string?>' doesn't match target type 'CL0<string>'.
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/Semantics/StaticNullChecking.cs]: >/ (26,9): warning CS8619: Nullability of reference types in value of type 'CL0<string?>' doesn't match target type 'CL0<string>'. [](start = 17, length = 129)

Looks like a dupe of the warning below.

File: src/Compilers/CSharp/Test/Semantic/Semantics/StaticNullChecking.cs [hunk 11]
```diff
@@ -8839,7 +8842,12 @@ class CL0<T>{}
 " }, parseOptions: TestOptions.Regular.WithNullCheckingFeature());
 
             c.VerifyDiagnostics(
-                );
+                // (13,17): warning CS8619: Nullability of reference types in value of type 'CL0<string?>' doesn't match target type 'CL0<string>'.
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/Semantics/StaticNullChecking.cs]: >// (13,17): warning CS8619: Nullability of reference types in value of type 'CL0<string?>' doesn't match target type 'CL0<string>'. [](start = 16, length = 131)

I would think that this is an explicit conversion, therefore should be no warnings.

File: src/Compilers/CSharp/Test/Semantic/Semantics/StaticNullChecking.cs [hunk 12]
```diff
@@ -14998,7 +15140,85 @@ static void G(string s)
             comp = CreateStandardCompilation(
                 new[] { source, attributesDefinitions },
                 parseOptions: TestOptions.Regular.WithNullCheckingFeature(flags: NullableReferenceFlags.IncludeNonNullWarnings).WithLanguageVersion(LanguageVersion.CSharp7_1));
-            comp.VerifyEmitDiagnostics();
+            comp.VerifyDiagnostics();
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/Semantics/StaticNullChecking.cs]: >VerifyDiagnostics [](start = 17, length = 17)

Why is this changed?

**Comment by AlekseyTs** [comment]
Done with review pass.

File: src/Compilers/CSharp/Portable/Lowering/DiagnosticsPass_ExpressionTrees.cs (hunk 7 above)
Comment (cston) [reply src/Compilers/CSharp/Portable/Lowering/DiagnosticsPass_ExpressionTrees.cs]: #21909

**Comment by gafter** [comment]
I don't know how to review this, as the [spec](https://github.com/dotnet/csharplang/blob/master/proposals/nullable-reference-types.md) is 7 months old and doesn't correspond with what is implemented here.

**Comment by gafter** [comment]
@cston What is the status of this PR?
